lib/linux_admin/network_interface.rb

```diff
@@ -0,0 +1,183 @@
+require 'ipaddr'
+
+module LinuxAdmin
+  class NetworkInterface
+    include Common
+
+    # Cached class instance variable for what distro we are running on
+    @dist_class = nil
+
+    # Gets the subclass specific to the local Linux distro
+    #
+    # @param clear_cache [Boolean] Determines if the cached value will be reevaluated
+    # @return [Class] The proper class to be used
+    def self.dist_class(clear_cache = false)
+      @dist_class = nil if clear_cache
+      @dist_class ||= begin
+        if [Distros.rhel, Distros.fedora].include?(Distros.local)
+          NetworkInterfaceRH
+        else
+          NetworkInterfaceGeneric
+        end
+      end
+    end
+
+    class << self
+      private
+
+      alias_method :orig_new, :new
+    end
+
+    # Creates an instance of the correct NetworkInterface subclass for the local distro
+    def self.new(*args)
+      if self == LinuxAdmin::NetworkInterface
+        dist_class.new(*args)
+      else
+        orig_new(*args)
```

Comment:
In this case it really isn't the super class new that we are looking for right?  This would be if someone called `.new` with one of the distro specific classes, then we would just call that class' `.new` method directly (which here is called `orig_new`).

Actually after thinking about this, maybe there is a better way to do this because the call stack here in the Red Hat case would look like:
NetworkInterfaceRH.new...
NetworkInterface.new line 36
NetworkInterface.new line 34
right?

So shouldn't this be:

```ruby
def self.new(*args)
  if self == LinuxAdmin::NetworkInterface
    dist_class.orig_new(*args)
  else
    orig_new(*args)
  end
end
```